File: rust/ceres-server/src/host.rs
//! The host interface.
//!
//! The server owns HTTP, but the engine it serves lives on the other side of this
//! trait: user lookup and credentials now, component operations and entity access as
//! the port grows. The production host is the Python engine reached through the
//! extension module's bridge, tests use stubs, and the boundary is exactly the seam a
//! future non-Python host would implement.

use std::path::PathBuf;

use axum::http::StatusCode;
use axum::response::{IntoResponse, Response};
use serde_json::Value;
use uuid::Uuid;

use crate::error::ApiError;

/// A user as the host reports it, enough for the actor gates plus its wire form.
pub struct UserRecord {
    pub id: Uuid,
    pub admin: bool,
    pub disabled: bool,
    /// The user's serialized wire form, password fields already excluded.
    pub payload: Value,
}

/// A failure crossing the host boundary.
///
/// A typed failure carries the status and serialized envelope the host produced, served
/// verbatim, anything else serves as a bare internal error like an unhandled exception
/// always has.
#[derive(Debug, thiserror::Error)]
pub enum HostError {
    #[error("the host reported a typed error")]
    Typed { status: u16, envelope: String },
    #[error("{0}")]
    Internal(String),
}

impl IntoResponse for HostError {
    fn into_response(self) -> Response {
        match self {
            Self::Typed { status, envelope } => {
                let status =
                    StatusCode::from_u16(status).unwrap_or(StatusCode::INTERNAL_SERVER_ERROR);
                (
                    status,
                    [(axum::http::header::CONTENT_TYPE, "application/json")],
                    envelope,
                )
                    .into_response()
            }
            Self::Internal(_) => {
                ApiError::new(StatusCode::INTERNAL_SERVER_ERROR, "error").into_response()
            }
        }
    }
}

/// A response the host described rather than serialized, whose body the server produces.
///
/// A procedure declaring a media type answers with one of these. The host decided every
/// header, including the content type and length, so the server sends them as they
/// arrive and only has to find the bytes.
pub struct Served {
    pub status: u16,
    /// The headers to send, in order, each name already lowercased by the host.
    pub headers: Vec<(String, String)>,
    /// The file to stream, or `None` when the body is the handle's chunks.
    pub file: Option<PathBuf>,
    /// The handle to release once the body ends, which runs the output's exit hook.
    pub handle: u64,
}

/// What an operation answered with.
pub enum Answer {
    /// A payload's verbatim JSON, which nearly every operation returns.
    ///
    /// The host already serialized it, so it flows into the response body untouched,
    /// a record dump never parses into a value tree on this side of the boundary.
    Payload(String),
    /// A body the server produces itself, which only a media output answers with.
    Served(Served),
}

/// Why a stream refused to open or stopped early.
#[derive(Debug)]
pub struct StreamClose {
    /// The WebSocket close code the socket reports.
    pub code: u16,
    /// The close reason, a serialized error for the codes that carry one.
    pub reason: String,
}

impl StreamClose {
    /// A refusal the engine did not attribute, reported as an internal error.
    pub fn internal(reason: impl Into<String>) -> Self {
        Self {
            code: 1011,
            reason: reason.into(),
        }
    }
}

/// A user's standing as the native gate reads it, no wire payload attached.
#[derive(Clone, Copy, Debug)]
pub struct GateUser {
    pub id: Uuid,
    pub admin: bool,
    pub disabled: bool,
}

/// The engine-side operations the server calls across the language boundary.
#[async_trait::async_trait]
pub trait Host: Send + Sync + 'static {
    /// Look up a user by ID, `None` when no such user exists.
    async fn user(&self, id: Uuid) -> Result<Option<UserRecord>, HostError>;

    /// Read a user's standing natively for an authentication gate.
    ///
    /// The outer `None` means this host has no native store and the caller falls back
    /// to the full `user` lookup. The inner `None` means no user carries the ID, which
    /// resolves to anonymous like every other token problem.
    async fn native_gate_user(&self, id: Uuid) -> Option<Option<GateUser>> {
        let _ = id;
        None
    }

    /// Serve a record listing natively from filter query pairs, as its response body.
    ///
    /// `None` delegates to the host operation, for an invalid filter whose canonical
    /// validation envelope the operation renders, a particle `class` filter, a host
    /// without a native store, or any native failure.
    async fn native_records(&self, table: &str, pairs: &[(String, String)]) -> Option<String> {
        let _ = (table, pairs);
        None
    }

    /// Count records natively from filter query pairs, as the count's JSON text.
    async fn native_record_count(&self, table: &str, pairs: &[(String, String)]) -> Option<String> {
        let _ = (table, pairs);
        None
    }

    /// Serve one record natively by ID, as its response body, `null` when absent.
    async fn native_record(&self, table: &str, id: Uuid) -> Option<String> {
        let _ = (table, id);
        None
    }

    /// Check a username and password, returning the user when they match.
    async fn verify_login(
        &self,
        username: String,
        password: String,
    ) -> Result<Option<UserRecord>, HostError>;

    /// Change a user's password, `None` when the old password does not match.
    ///
    /// The host validates the new password's shape, reporting refusals as typed
    /// errors.
    async fn change_password(
        &self,
        user: Uuid,
        old_password: String,
        new_password: String,
    ) -> Result<Option<UserRecord>, HostError>;

    /// Open a stream, answering with the handle its messages arrive under.
    ///
    /// A refusal carries the close code the socket reports, so the policy for which
    /// failure closes with which code stays with the engine.
    async fn stream_open(&self, operation: &str, arguments: Value) -> Result<u64, StreamClose> {
        let _ = arguments;
        Err(StreamClose::internal(format!(
            "this host does not stream {operation:?}"
        )))
    }

    /// Await the next message on a stream, `None` once it ends.
    ///
    /// Messages arrive pre-serialized, so a record the engine already rendered crosses
    /// the boundary once as text.
    async fn stream_next(&self, handle: u64) -> Result<Option<String>, StreamClose> {
        let _ = handle;
        Ok(None)
    }

    /// Await the next chunk of a described response's body, `None` once it ends.
    ///
    /// Chunks arrive as raw bytes, so a body the host produces crosses without an
    /// encoding step of its own.
    async fn next_chunk(&self, handle: u64) -> Result<Option<Vec<u8>>, HostError> {
        let _ = handle;
        Ok(None)
    }

    /// Release a stream's resources, whatever ended it.
    async fn stream_close(&self, handle: u64) {
        let _ = handle;
    }

    /// Run a named engine operation, the generic channel most route families ride.
    ///
    /// The operation names and argument shapes form the contract between the server and
    /// its host, one name per route behavior, and the answer is what the route serves.
    async fn operate(&self, operation: &str, arguments: Value) -> Result<Answer, HostError> {
        let _ = arguments;
        Err(HostError::Internal(format!(
            "this host does not support the {operation:?} operation"
        )))
    }

    /// Run an operation that answers with a payload.
    ///
    /// Only the procedure call routes reach an operation that can describe a response,
    /// so every other route asks for the payload and treats a description as a failure
    /// of the host rather than carrying a branch that cannot be taken.
    async fn payload(&self, operation: &str, arguments: Value) -> Result<String, HostError> {
        match self.operate(operation, arguments).await? {
            Answer::Payload(payload) => Ok(payload),
            Answer::Served(_) => Err(HostError::Internal(format!(
                "the {operation:?} operation answered with a body of its own"
            ))),
        }
    }
}

/// A host for applications that never cross the boundary, tests and stubs.
pub struct NoHost;

#[async_trait::async_trait]
impl Host for NoHost {
    async fn user(&self, _id: Uuid) -> Result<Option<UserRecord>, HostError> {
        Ok(None)
    }

    async fn verify_login(
        &self,
        _username: String,
        _password: String,
    ) -> Result<Option<UserRecord>, HostError> {
        Ok(None)
    }

    async fn change_password(
        &self,
        _user: Uuid,
        _old_password: String,
        _new_password: String,
    ) -> Result<Option<UserRecord>, HostError> {
        Ok(None)
    }
}

File: ceres/__internal__/app/host.py
"""The engine as the native server's host.

The native server owns HTTP and answers most requests by calling one named operation
here. Each operation validates its own arguments through Pydantic and runs the same
engine and query-layer code the API always has, so filters, permissions, and wire
shapes keep their exact behavior while the transport moves to Rust.

Results cross the boundary as one JSON envelope per call, `{"ok": ...}` for a payload
and `{"error": {"status", "envelope"}}` for a typed refusal, which the server serves
verbatim. Streams open under a handle and yield one pre-serialized message at a time.
"""

from __future__ import annotations

import json
from collections.abc import AsyncIterator, Callable
from itertools import count
from typing import TYPE_CHECKING, Any

from pydantic import ValidationError

from ceres.data import to_json, validate
from ceres.error import (
    Error,
    HTTPError,
    NotAuthenticatedError,
    NotFoundError,
    ValidationFailedError,
    ValidationProblem,
)
from ceres.user import User

if TYPE_CHECKING:
    from uuid import UUID

    from ceres.engine import Engine

_OPERATIONS: dict[str, Callable[[Host, dict[str, Any]], Any]] = {}
"""Every named operation, resolved once at import."""

_STREAMS: dict[str, Callable[[Host, dict[str, Any]], AsyncIterator[str]]] = {}
"""Every named stream, each an async iterator of serialized messages."""


def operation(name: str):
    """Register an async operation under its name."""

    def register[T: Callable[..., Any]](function: T) -> T:
        _OPERATIONS[name] = function
        return function

    return register


def stream(name: str):
    """Register an async generator of serialized messages under its name."""

    def register[T: Callable[..., Any]](function: T) -> T:
        _STREAMS[name] = function
        return function

    return register


class Host:
    """The engine behind the native server's host interface."""

    def __init__(self, engine: Engine) -> None:
        self._engine = engine
        self._streams: dict[int, AsyncIterator[str]] = {}
        self._handles = count(1)

    @property
    def engine(self) -> Engine:
        return self._engine

    async def user(self, id: UUID) -> str:
        """Look up a user by ID for the server's identity resolution."""
        return _record(await self._engine.users.get(id))

    async def verify_login(self, username: str, password: str) -> str:
        """Check a username and password, answering with the user when they match."""
        user = await self._engine.users.where(username=username).first()
        if user is None or not await self._engine.verify_password(password, user.password):
            return _record(None)

        return _record(user)

    async def change_password(self, id: UUID, old_password: str, new_password: str) -> str:
        """Replace a user's password, refusing when the old one does not match."""
        try:
            user = await self._engine.users.get(id)
            if user is None or not await self._engine.verify_password(old_password, user.password):
                return _record(None)

            await self._engine.users.where(id=id).update(
                validate(User.Update, {"password": new_password})
            )
            return _record(await self._engine.users.get(id))
        except Error as error:
            return _failure(error)

    async def operate(self, name: str, arguments: str) -> str:
        """Run one named operation, answering with its result envelope."""
        try:
            handler = _OPERATIONS.get(name)
            if handler is None:
                raise NotFoundError()

            payload = await handler(self, json.loads(arguments))
            return json.dumps({"ok": payload})
        except Error as error:
            return _failure(error)
        except ValidationError as error:
            # An operation validates its own arguments, so a validation failure is a bad
            # request rather than an internal one, reported with its problems like the
            # framework's own handler always did.
            return _failure(ValidationFailedError(problems=ValidationProblem.extract(error)))
        except Exception as error:  # noqa: BLE001
            del error
            return _failure(HTTPError(status=500))

    async def stream_open(self, name: str, arguments: str) -> str:
        """Open a stream, answering with the handle its messages arrive under."""
        try:
            opener = _STREAMS.get(name)
            if opener is None:
                raise NotFoundError()

            iterator = opener(self, json.loads(arguments))
            handle = next(self._handles)
            self._streams[handle] = iterator
            return json.dumps({"ok": handle})
        except ValidationError as error:
            return json.dumps(
                {
                    "close": _close_of(
                        ValidationFailedError(problems=ValidationProblem.extract(error))
                    )
                }
            )
        except Error as error:
            return json.dumps({"close": _close_of(error)})
        except Exception as error:  # noqa: BLE001
            return json.dumps({"close": {"code": 1011, "reason": str(error)[:120]}})

    async def stream_next(self, handle: int) -> str:
        """Await the next message on a stream, reporting its end or its failure."""
        iterator = self._streams.get(handle)
        if iterator is None:
            return json.dumps({"end": True})

        try:
            return json.dumps({"message": await anext(iterator)})
        except StopAsyncIteration:
            return json.dumps({"end": True})
        except Error as error:
            return json.dumps({"close": _close_of(error)})
        except Exception as error:  # noqa: BLE001
            return json.dumps({"close": {"code": 1011, "reason": str(error)[:120]}})

    async def stream_close(self, handle: int) -> None:
        """Release a stream, closing its iterator if it supports closing."""
        iterator = self._streams.pop(handle, None)
        closer = getattr(iterator, "aclose", None)
        if closer is not None:
            try:
                await closer()
            except Exception:  # noqa: BLE001, S110
                pass


def _record(user: User | None) -> str:
    """The envelope carrying a user record, or its absence."""
    if user is None:
        return json.dumps({"ok": None})

    payload = json.loads(to_json(user, exclude={"password": True}))
    return json.dumps(
        {
            "ok": {
                "id": str(user.id),
                "admin": user.admin,
                "disabled": user.disabled,
                "payload": payload,
            }
        }
    )


def _failure(error: Error) -> str:
    """The envelope carrying a typed refusal, with its status and serialized form."""
    from ceres.error import simplify

    return json.dumps(
        {
            "error": {
                "status": error.__error_status_code__,
                "envelope": json.loads(to_json(simplify(error))),
            }
        }
    )


def _close_of(error: Error) -> dict[str, Any]:
    """The close code and reason a stream failure reports.

    A refusal the caller could have avoided closes with a policy violation carrying the
    error, and anything else reports an internal failure, matching the codes the socket
    routes have always sent.
    """
    from ceres.error import ProcedureError, ProcedureInternalError, simplify

    policy = isinstance(error, NotFoundError | NotAuthenticatedError | ProcedureInternalError)
    if not policy and isinstance(error, ProcedureError):
        code = 1011
    else:
        code = 1008 if policy else 1011

    return {"code": code, "reason": to_json(simplify(error))}
